### PR TITLE
Filter out diff deltas that don't match the spatial filter.

### PR DESCRIPTION
**Pull request by olsen232**
## Description

Feature deltas are only relevant to the user if either the old or the new version of the feature is inside the spatial filter. The exception to this rule is local working-copy edits, which are always relevant to the user. This PR filters out irrelevant deltas that do not match the spatial filter.

## Related links:

https://github.com/koordinates/kart/issues/456

### CODE REVIEW

**Review by rcoup — APPROVED** [review]
> Spatial filtered diff output is not yet properly streamed - every feature is loaded before the first feature is output. (In contrast to diffs where there is no spatial filter, where each feature is loaded just in time). This should be fixed at the very least for json-lines output, which is where streaming is most beneficial.

Is there a reason that yielding matching features one at a time isn't straightforward? Or simply to get it going?

**Comment by olsen232** [comment]
> > Spatial filtered diff output is not yet properly streamed - every feature is loaded before the first feature is output. (In contrast to diffs where there is no spatial filter, where each feature is loaded just in time). This should be fixed at the very least for json-lines output, which is where streaming is most beneficial.
> 
> Is there a reason that yielding matching features one at a time isn't straightforward? Or simply to get it going?

I've done it now, so, it's not impossible - it just took a bit of restructuring. As the code was, filtering feature deltas immediately before they were output was not possible, since at that point, there was no way to tell which deltas were from commits and which were from the WC (and the two sources are filtered differently). However, filtering feature deltas at the point where we did have this information was too soon - this information was lost before the first feature was output, so if we do spatial filtering at that point, it means loading every feature before outputting the first one.

The fix is just to propagate which features are from commits and which features are from the WC.

Also added a warning to `kart create-patch` so that TODO is done now too - we can always revisit this if we decide `kart create-patch` should not have be affected by spatial filtering at all.

File: kart/base_diff_writer.py [hunk 1]
```diff
@@ -237,54 +245,63 @@ def get_dataset_diff(self, dataset_path, apply_spatial_filter=True, prune=True):
...
-        Removes all the feature-deltas from the given dataset diff that don't match the self.spatial_filter.
-        Features that are present in the working-copy diff for this dataset - if supplied - are not
-        removed from the diff, since working-copy edits are always relevant to the user even if outside the
-        spatial filter.
+        Yields the key, delta for only those feature-deltas from the given dataset diff that match
+        self.spatial_filter. Note that feature-deltas are always considered to match the spatial-filter
+        if they are marked as working-copy edits, since working-copy edits are always relevant to the user
+        even if they are outside the spatial filter.
         """
         # NOTE: This function has to load every feature if it is to do any filtering at all.
```
Comment (rcoup) [inline kart/base_diff_writer.py:266]: Comments not relevant anymore?

Comment (olsen232) [reply kart/base_diff_writer.py:266]: Removed

**Pull request merged**